ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -1947,30 +1950,70 @@ prop_diffusion_async_demotions defaultBearerInfo diffScript =
                        _            -> False)
             . Signal.fromEventsWith (Right Set.empty)
             . Signal.selectEvents
-                (\case DiffusionPeerSelectionActionsTrace (PeerStatusChanged (HotToCold connId)) ->
+                (\case DiffusionPeerSelectionActionsTrace (PeerStatusChanged (HotToCooling connId)) ->
                            Just $ Right demotions
                          where
                            demotions = Set.singleton (remoteAddress connId)
+                       DiffusionPeerSelectionActionsTrace (PeerStatusChanged (WarmToCooling connId)) ->
+                           Just $ Right demotions
+                         where
+                           demotions = Set.singleton (remoteAddress connId)
+                       DiffusionConnectionManagerTrace (TrConnectionCleanup connId) ->
+                           Just $ Left failures
+                         where
+                           failures = Just $ Set.singleton (remoteAddress connId)
                        DiffusionPeerSelectionTrace (TraceDemoteAsynchronous status) ->
                            Just $ Left (Just failures)
                          where
-                           failures = Map.keysSet (Map.filter ((==PeerCold) . fst) status)
+                           failures = Map.keysSet (Map.filter ((==PeerCooling) . fst) status)
```

Comment:
I am not sure if this is right. Asynchronous demotion will bring connection down immediately without going through the `PeerCooling` step (check the `cleanup` handler in `forkConnectionHandler` in `ConnectionManager.Core`).

Reply:
I think that's right because when an exception happens you set the state of the connection to `PeerCooling` in `peerMonitoringLoop`. But if there was an exception, the connection manager will shutdown the connection without going through the `WAITTIME` interval (which `PeerCooling` models on the `PeerStateActions` side).